### ISSUE TITLE
ASV Benchmarking Figure

### ISSUE THREAD

**Issue by mdhaber** [scipy/scipy-articles#143]
There have been concerns about the quality of the ASV Benchmarking Figure (#65, #117).

![asv_time_query_ckdtree](https://user-images.githubusercontent.com/6570539/57961177-0b4b8d00-78c2-11e9-8846-f8253cf723b6.png)

Apparently, the data are no longer available and are difficult to generate, so I scraped the data from [here](https://pv.github.io/scipy-bench/#spatial.CNeighbors.time_count_neighbors_deep?y-axis-scale=log&p-Nr=10&p-Nr=100&p-Nr=1000&p-(m%2C%20n1%2C%20n2)=(2%2C%201000%2C%201000)&p-(m%2C%20n1%2C%20n2)=(8%2C%201000%2C%201000)&p-(m%2C%20n1%2C%20n2)=(16%2C%201000%2C%201000)&x-axis-scale=date%2Finfo.json&Cython=0.27.3&python=2.7&numpy=1.13.3) for `spatial.CNeighbors.time_count_neighbors_deep` benchmark and produced:

![asv](https://user-images.githubusercontent.com/6570539/57961153-e7884700-78c1-11e9-9a16-fd6c7294f9d4.png)

If you have style tweak suggestions that would be easy for me to implement (in Inkscape), I will consider them. I would welcome suggestions (as RGBA hex codes) for the color palette. I would also welcome a draft caption, as I don't know what this benchmark really does.

I will submit a PR to replace the figure after I find out what the improvements were. If you happen to have ideas about those, please let me know. In case it helps, the dates and commit hashes corresponding with each data point are [here](https://www.dropbox.com/s/enw9hy6g9v81344/asvdata.csv?dl=0).

I have no attachment to this benchmark and I'd prefer to have a denser set of points, so if the data for the original benchmark is found or recreated I'd be happy to swap it in.

**Comment by tylerjereddy** [comment]
I'm -1 here.

I don't think we should be changing what we benchmark, the text surrounding the benchmark section, or the figure caption at this stage, simply because there is some work to regenerating the data. Considerable effort went into selecting an appropriate benchmark, and generating the original benchmarks, with a config script now in our supporting information, and substantial discussion with @stsievert. I'm pretty sure the two of us agreed that the screen capture was "ok" if not ideal for its representation of what you actually see when you run it.

I'd say 3 people should speak up for action on that figure so there's a majority.

If we need a high resolution version at submission time then I can try to regenerate the data & provide the additional config scripts from multiple passes with different dependencies to fill in build failures, etc. to improve prospects for future reproducible work. 

Really, dependency matrices for fallbacks would be useful in asv, which was also discussed previously while making this figure, but without those a bit of manual tweaking is needed for runs on ancient commit hashes.

**Comment by ilayn** [comment]
I think this screenshot packed with dots conveys no information and to be honest I think nobody cares. We know what it does and readers won't get anything out of such cockpit view. We might want to collect a few functions with drastic improvements over the years and show that we keep track of performance of things and respective regressions. But as is, it would just waste precious space (even with the hi-res version) if we don't have a clear message.

**Comment by rgommers** [comment]
> I'd say 3 people should speak up for action on that figure so there's a majority.

Sorry I didn't find the bandwidth to comment before. I think @mdhaber and @ilayn are right though. Either we remove it, or we simply replace it with @mdhaber's version. The original has text that will be unreadably small, and therefore not accepted for publication.

**Comment by mdhaber** [comment]
@tylerjereddy I'm -1 for changing the benchmark, too. I tried to regenerate the data, but the command that was in the text of the paper didn't work for me. I even tried to digitize the data from the graph, but I didn't feel comfortable with that because I was having to manually impute some data. At the time, you didn't seem keen on reproducing the data, so I thought this was the next best option. But as I mentioned in the original post, I'd be happy to swap in the data for the original benchmark if it is recovered or recreated.

@ilayn This was in the key technical improvements section. The intent of the figure, presumably, was to reinforce the message from the text that performance is tracked and that cKDTree has improved. Let me try writing a caption and putting it in context so we can better decide whether the message is strong enough to warrant the use of space.

**Comment by tylerjereddy** [comment]
> but the command that was in the text of the paper didn't work for me.

Can you be more specific? `asv` has pretty detailed docs--normally it will produce a traceback when it fails, and then you just kind of keep grinding away until all the dependencies and requirements are satisfied.

From there, you keep re-running with different versions of Cython and NumPy until a complete set of data is regenerated without major gaps in history.

**Comment by tylerjereddy** [comment]
> Either we remove it, or we simply replace it

Or I regenerate the data, and then no text surrounding it has to change.

**Comment by mdhaber** [comment]
@tylerjereddy That would be great. I asked in #117:
> If you can send the raw data, I can draft a figure.

You suggested:
> if there's a request for revision then we can suffer through the regeneration of data if needed.

I didn't feel comfortable sending this out for review as it was, though, so I did my best to replace it without troubling you further. I failed again : (

I don't remember the specific difficulties I experienced trying to reproduce the data. I spent a few hours grinding, but at some point I hit a major roadblock and chose to change course rather than trying to push through it.

I agree that if we could produce a figure with the original data, none of the surrounding text would have to change. Only the caption would need to be tweaked, as the reader probably doesn't know what the $L^2$ norm, toroidal topology, or `boxsize` argument have to do with the test.  Also, it sounded like you wanted to add the console command back in, and the caption wouldn't be a bad place for it.

**Comment by ilayn** [comment]
>This was in the key technical improvements section. The intent of the figure, presumably, was to reinforce the message from the text that performance is tracked and that cKDTree has improved. Let me try writing a caption and putting it in context so we can better decide whether the message is strong enough to warrant the use of space.
Yes that's perfectly fine let me rephrase in case it came out wrong; the figure is not conveying the message about the amount of work SciPy team put in just to make things more performant. If we can pull the data out somehow (please let me know if I can be of use including manually reading off the dots), we can make a more stronger point.

**Comment by tylerjereddy** [comment]
I'm starting the asv runs again locally.

I kept very detailed notes of the procedure I used the first time around: https://workflowy.com/s/im-gradually-refinin/13IzABVs7tJftdmA

Here are the working notes on the regeneration procedure: https://workflowy.com/s/i-need-to-regenerate/yug1dImVfF6C8FTu

**Comment by mdhaber** [comment]
@ilayn right, I think I agree. The dots and lines alone don't tell the story. Note the markers I added in the new version of the figure - "this happened"/"that happened". These are placeholders for notes about what was changed in the code that led to the observed drop in execution time. For instance, when the data is finished being regenerated and I remake this figure, I would note that the first drop is due to Cythonizing important functions and the second drop is due to the C++ rewrite. Do you agree that this - along with a strong caption - can make a strong point?

**Comment by ilayn** [comment]
Yes I think that would be a stronger point.

**Comment by mdhaber** [comment]
Here is a new version with the data from #144.

![asv](https://user-images.githubusercontent.com/6570539/58842841-ce52fa80-8624-11e9-87f8-822163f1f91f.png)

If we like this, I plan to rescale to reduce vertical space and submit a PR. 
Thoughts?

**Comment by mdhaber** [comment]
Caption: Results of the `scipy.spatial.cKDTree.query` benchmark from the introduction of `cKDTree` to the release of SciPy 1.0. The benchmark generates a _k_-d tree from 10,000 uniformly distributed points in an _m_-dimensional unit hypercube, then finds the nearest (Euclidean) neighbor in the tree for each of 1,000 query points. Each marker in the figure indicates the execution time of the benchmark for a commit in the master branch of SciPy. The tree was generated without application of toroidal topology (`boxsize=None`), and tests were performed by Airspeed Velocity 0.4 using Python 2.7, NumPy 1.8.2, and Cython versions 0.27.3, 0.21.1, and 0.18 (for improved backward
compatibility).

**Comment by rgommers** [comment]
Looks great now, thanks @tylerjereddy and @mdhaber! 

The "rewritten in C++" vertical line should be moved a little to the right.

@mdhaber question: did you go through the commit history and verify those two rewrites where indeed the causes? Or did you rely on my imperfect memory of events?

**Comment by ilayn** [comment]
> The "rewritten in C++" vertical line should be moved a little to the right.

Aha here we go 😃  -> [obligatory oat meal](https://theoatmeal.com/comics/design_hell)

**Comment by mdhaber** [comment]
@rgommers I checked roughly based on the release candidate numbers of the old plot, but now that I have the actual commit hashes I'll check those.
Re:vertical lines You mean like a single pixel so it goes through the center of the dots, right?

**Comment by ilayn** [comment]
I think he means the verticals are aligned with the performance gain drops

**Comment by mdhaber** [comment]
@ilayn re:horizontal grid there are tick marks. Want them a little longer, or faint lines across the whole plot?

**Comment by ilayn** [comment]
Ah now you mention it, I see them. Thanks

**Comment by stsievert** [comment]
> re:horizontal grid there are tick marks. Want them a little longer, or faint lines across the whole plot?

I think I'd like to see the faint horizontal grid lines to help with comparison. I'm trying to see how much cKDTree improved between jumps, but can't tell. I think faint grid lines would help with "cKDTree speed improved by 2x after fully Cythonizing".

**Comment by mdhaber** [comment]
Better or worse?

![asv](https://user-images.githubusercontent.com/6570539/58993692-17c95400-87a3-11e9-9534-2d3de28cabe9.png)

**Comment by stsievert** [comment]
That’s an effective resolution of my style nit – the comparisons are easy visualize now. Thanks!

**Comment by mdhaber** [comment]
I also wanted to squash it vertically.
![asv](https://user-images.githubusercontent.com/6570539/58997357-17d05080-87b1-11e9-950f-3e6fdb28c69d.png)
I'll submit a PR with this one.

**Comment by mdhaber** [comment]
@rgommers The commit corresponding with the line labeled `cKDTree` fully cythonized is [b9370d977230176f0ec7106fbb5f8009c124271f](https://github.com/scipy/scipy/commit/b9370d977230176f0ec7106fbb5f8009c124271f). I'm not sure what to make of that.

**Comment by mdhaber** [comment]
There is a ton of work on cKDTree around that date:
https://github.com/scipy/scipy/commits/master?before=b9370d977230176f0ec7106fbb5f8009c124271f+221

It's parent's commit date is several months in the future.

**Comment by mdhaber** [comment]
The next commit with the lower execution time is:
[a5734f4f2fa5da181bcf07fecaa95edf5a6ae393](https://github.com/scipy/scipy/commit/a5734f4f2fa5da181bcf07fecaa95edf5a6ae393)

One of its grandparents is
[7c6796dc0db936ab4f85cda7fdf5c2b5a65f6bd3](https://github.com/scipy/scipy/commit/7c6796dc0db936ab4f85cda7fdf5c2b5a65f6bd3)

which was the last in a series of cKDTree improvements by @patvarilly, adding to the release notes:
```
cKDTRee improvements
--------------------
scipy.spatial: Cython version of KDTree, cKDTree, is now feature-complete.
Most operations (construction, query, query_ball_point, query_pairs,
count_neighbors and sparse_distance_matrix) are between 200 and 1000 times
faster in cKDTree than in KDTree.
```

I think the only way to be more sure that the Cythonization was responsible for the change would be to benchmark _every_ commit around that time and see when the change occurred. I don't think it's necessary. In any case, the plot itself doesn't say that Cythonizing `cKDTree` is what caused the change. It just indicates when the Cythonization happened. That is when the Cythonization happened.

**Comment by rgommers** [comment]
sounds reasonable, agree no need to generate more data

**Comment by mdhaber** [comment]
Now for the conversion to C++, again the first commit with the lower time is:
[https://github.com/scipy/scipy/commit/8d717dd393dc6c6da422ebae49e208db4ce20be2](8d717dd393dc6c6da422ebae49e208db4ce20be2)
Again, its parent is far in the future, after the execution time of all the commits has dropped.

The next commit with the lower time is:
[3bb02c0fc24cc7fc852dd260ccfcdb76c8878ab2](https://github.com/scipy/scipy/commit/3bb02c0fc24cc7fc852dd260ccfcdb76c8878ab2)
One of its parents is cKDTree work, and a few parents back it's the [0.17 release notes](https://docs.scipy.org/doc/scipy/reference/release.0.17.0.html#scipy-spatial-improvements).
